File: tmdb_cache/scrape.py
import requests
import os
import json
import time

from dotenv import load_dotenv
from urllib.parse import urlencode
from pprint import pprint

load_dotenv()

API_FAILURES_LOG = "api_failures.json"
API_URL = "https://api.themoviedb.org/3"
API_TOKEN = os.environ.get("TMBD_API_TOKEN")
HEADERS = dict(
    {
        "accept": "application/json",
        "Authorization": f"Bearer {API_TOKEN}",
    },
)

# Load cache in memory
CACHE = {}
start = time.time()
for directory in ["./person", "./movie"]:
    for filename in os.listdir(directory):
        key = "/".join([directory.split("/")[1], filename.split(".")[0]])
        with open(f"{directory}/{filename}", "r") as f:
            CACHE[key] = json.load(f)
end = time.time()
print(f"Loaded cached API responses in {end - start}s")

with open(API_FAILURES_LOG, "r") as f:
    API_FAILURES = json.load(f)


def key_to_filepath(key):
    parts = key.split("/")
    directory = parts[0]
    rest = "_".join(parts[1:])
    return f"./{directory}/{rest}.json"


def fetch_and_cache(url_key):
    parts = url_key.split("/")
    cache_key = "/".join([parts[0], "_".join(parts[1:])])
    if url_key in API_FAILURES and API_FAILURES[url_key] >= 5:
        return None
    if cache_key in CACHE:
        print(f"Cache hit for {url_key}")
        return CACHE[cache_key]
    try:
        print(f"Cache miss for {url_key}")
        response = requests.get(f"{API_URL}/{url_key}", headers=HEADERS, timeout=5)
        data = json.loads(response.content)
        CACHE[cache_key] = data
        with open(key_to_filepath(url_key), "w") as f:
            json.dump(data, f)
        return data
    except:
        print(f"Failed fetching {url_key} from TMDB API")
        API_FAILURES[url_key] = API_FAILURES.get(url_key, 0) + 1


def search_title(title):
    qs = urlencode({"query": title})
    key = f"search/movie?{qs}"
    return fetch_and_cache(key)


def person_credits(id):
    return fetch_and_cache(f"person/{id}/movie_credits")


def movie_credits(id):
    return fetch_and_cache(f"movie/{id}/credits")


def find_director(crew):
    for item in crew:
        if item["department"] == "Directing" and item["job"] == "Director":
            return item


def parse_credits(credits, limit=None):
    if not credits:
        return []
    cast = credits.get("cast", [])
    director = find_director(credits.get("crew", []))
    results = [director] + cast
    if not limit:
        return list(filter(lambda item: item, results))
    return list(filter(lambda item: item, results))[:limit]


# List of people who directed or acted in movie
def cast_and_director(movie_id, limit=None):
    return parse_credits(movie_credits(movie_id), limit=limit)


# List of movies person directed or acted in
def acting_directing_credits(person_id, limit=None):
    return parse_credits(person_credits(person_id), limit=limit)


def crawl(movie_id=None, person_id=None, depth=1, limit=None):
    if depth < 1:
        return
    elif not movie_id and not person_id:
        print("crawl called without args")
        return
    elif movie_id:
        people = cast_and_director(movie_id, limit=limit)
        for person in people:
            crawl(person_id=person["id"], depth=depth - 1)
    elif person_id:
        movies = acting_directing_credits(person_id, limit=limit)
        for movie in movies:
            crawl(movie_id=movie["id"], depth=depth - 1)


# print("Uncomment line to scrape!")
crawl(person_id=2150101, depth=4, limit=5)

with open(API_FAILURES_LOG, "w") as f:
    json.dump(API_FAILURES, f)

File: scripts/scrape.js
#!/usr/bin/env node

const fs = require("fs-extra");
const { readFile, writeFile } = require("fs/promises");
const _ = require("lodash");
const fetch = require("node-fetch");
const { RateLimit } = require("async-sema");
const { program } = require("commander");

require("dotenv").config();

program
  .option("-d, --depth <number>", "crawl depth", parseInt)
  .option("-m, --movie <string>", "movie id")
  .option("-p, --person <string>", "person id");

program.parse(process.argv);

const options = program.opts();

const API_URL = "https://api.themoviedb.org/3";
const HEADERS = {
  accept: "application/json",
  Authorization: `Bearer ${process.env.TMBD_API_TOKEN}`,
};
const MOVIE_API_RESPONSES_DIRECTORY = "./static/movie";
const PERSON_API_RESPONSES_DIRECTORY = "./static/person";

const CACHED_API_RESPONSES = new Set([
  ...fs.readdirSync(MOVIE_API_RESPONSES_DIRECTORY),
  ...fs.readdirSync(PERSON_API_RESPONSES_DIRECTORY),
]);

const throttle = RateLimit(3);

const readFromCache = (directory, filename) =>
  readFile(`${directory}/${filename}`, "utf8").then((data) => JSON.parse(data));

const writeToCache = (data, directory, filename) =>
  writeFile(`${directory}/${filename}`, JSON.stringify(data), "utf8").then((_) => {
    CACHED_API_RESPONSES.add(filename);
    return data;
  });

async function fetchAndCache(urlKey, directory, filename) {
  if (CACHED_API_RESPONSES.has(filename)) {
    console.log(`HIT for ${filename}`);
    return readFromCache(directory, filename);
  }
  await throttle();
  if (CACHED_API_RESPONSES.has(filename)) {
    console.log(`HIT for ${filename}`);
    return readFromCache(directory, filename);
  }
  console.log(`MISS for ${filename}, fetching from API`);
  return fetch(`${API_URL}/${urlKey}`, {
    method: "GET",
    headers: HEADERS,
    timeout: 3000,
  })
    .then((response) => {
      if (response.ok) return response.json();
      else {
        return response.json().then((data) => {
          throw _.get(data, "status_message", "Unknown error");
        });
      }
    })
    .then((data) => writeToCache(data, directory, filename))
    .catch((e) => console.log(`Error fetching ${urlKey}: ${e}`));
}

const movieCredits = (movieId) =>
  fetchAndCache(
    `movie/${movieId}/credits`,
    MOVIE_API_RESPONSES_DIRECTORY,
    `${movieId}_credits.json`,
  );

const personCredits = (personId) =>
  fetchAndCache(
    `person/${personId}/movie_credits`,
    PERSON_API_RESPONSES_DIRECTORY,
    `${personId}_movie_credits.json`,
  );

const principals = (credits) => [
  ..._.get(credits, "crew", []).filter((credit) => _.get(credit, "job") === "Director"),
  ..._.get(credits, "cast", []),
];

const scrape = (movieId = null, personId = null, depth = 1) => {
  if (depth < 1 || (!movieId && !personId)) return;
  if (movieId) {
    movieCredits(movieId)
      .then(principals)
      .then((people) => people.map((person) => person.id))
      .then((personIds) => personIds.forEach((personId) => scrape(null, personId, depth - 1)));
  } else if (personId) {
    personCredits(personId)
      .then(principals)
      .then((movies) => movies.map((movie) => movie.id))
      .then((movieIds) => movieIds.forEach((movieId) => scrape(movieId, null, depth - 1)));
  }
};

const depth = Math.min(options.depth, 4);
console.log(
  `Crawling ${
    options.movie ? `Movie ${options.movie}` : `Person ${options.person}`
  } to depth ${depth}`,
);
scrape(options.movie, options.person, Math.min(options.depth, 4) || 1);
